src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -16,168 +17,383 @@
+		else
+		{
+			layouts[texturesInfo.setIx] = utilities->getLogicalDevice()->createDescriptorSetLayout({ {textureBinding} });
+			layouts[samplersInfo.setIx] = utilities->getLogicalDevice()->createDescriptorSetLayout({ {samplersBinding} });
+		}
+
+		assert(layouts[texturesInfo.setIx]);
+		assert(layouts[samplersInfo.setIx]);
+
+		return utilities->getLogicalDevice()->createPipelineLayout(PushConstantRanges, std::move(layouts[0u]), std::move(layouts[1u]), std::move(layouts[2u]), std::move(layouts[3u]));
+	}
+
+	const smart_refctd_ptr<IFileArchive> UI::mount(smart_refctd_ptr<ILogger> logger, ISystem* system, const std::string_view archiveAlias)
+	{
+		assert(system);
+		
+		if(!system)
+			return nullptr;
+
+		auto archive = make_smart_refctd_ptr<builtin::CArchive>(smart_refctd_ptr(logger));
+		system->mount(smart_refctd_ptr(archive), archiveAlias.data());
+
+		return smart_refctd_ptr(archive);
+	}
+
+	void UI::createPipeline(SCreationParameters& creationParams)
+	{
+		auto pipelineLayout = smart_refctd_ptr<IGPUPipelineLayout>(creationParams.pipelineLayout);
 
 		if (!pipelineLayout)
 		{
-			m_creationParams.utilities->getLogger()->log("Could not create pipeline layout!", system::ILogger::ELL_ERROR);
+			creationParams.utilities->getLogger()->log("Could not create pipeline layout!", ILogger::ELL_ERROR);
 			assert(false);
 		}
 
 		struct
 		{
-			core::smart_refctd_ptr<video::IGPUShader> vertex, fragment;
+			smart_refctd_ptr<IGPUShader> vertex, fragment;
 		} shaders;
 
 		{
 			constexpr std::string_view NBL_ARCHIVE_ALIAS = "nbl/ext/imgui/shaders";
 				
-			auto system = smart_refctd_ptr<system::ISystem>(m_creationParams.assetManager->getSystem());																	//! proxy the system, we will touch it gently
-			auto archive = make_smart_refctd_ptr<nbl::ext::imgui::builtin::CArchive>(smart_refctd_ptr<system::ILogger>(m_creationParams.utilities->getLogger()));		//! we should never assume user will mount our internal archive since its the extension and not user's job to do it, hence we mount only to compile our extension sources then unmount the archive
-			auto compiler = make_smart_refctd_ptr<CHLSLCompiler>(smart_refctd_ptr(system));																				//! note we are out of default logical device's compiler set scope so also a few special steps are required to compile our extension shaders to SPIRV
+			//! proxy the system, we will touch it gently
+			auto system = smart_refctd_ptr<ISystem>(creationParams.assetManager->getSystem());
+
+			//! note we are out of default logical device's compiler set scope so also a few special steps are required to compile our extension shaders to SPIRV
+			auto compiler = make_smart_refctd_ptr<CHLSLCompiler>(smart_refctd_ptr(system));	
 			auto includeFinder = make_smart_refctd_ptr<IShaderCompiler::CIncludeFinder>(smart_refctd_ptr(system));
 			auto includeLoader = includeFinder->getDefaultFileSystemLoader();
 			includeFinder->addSearchPath(NBL_ARCHIVE_ALIAS.data(), includeLoader);
```

Comment:
wouldn't just mounting the extension builtin resources under the correct alias OR creating a folder->folder alias (mount folder) completely eliminate the need for this?

Reply:
> under the correct alias 

the only reason for me to give alias here is to use asset manager's CWD to load shaders easily (short asset keys, just `vertex.hlsl & fragment.hlsl`)

Reply:
Also there's a compiler in AssetManager, and a compiler set in ILogicalDevice, and IIRC they have different include finders.

Its best to rely on exisitng search paths (`include`) and include headers by full paths so that:
1. you don't cause path clashes with user/other extension shaders
2. the shaders can compile outside of your extension without having to add extra include paths

Reply:
> but because you've added the search paths

I created `includeFinder` in the scope of this call

> there's a vertex.hlsl or fragment.hlsl the include finder might find the imgui ones.

it won't because:

- it dies as soon as I exit the scope
- as soon as shaders are compiled the extension archive is unmounted

Reply:
what's the rationale for unmounting the archive?

Reply:
> IIRC they have different include finders.

yes correct

> compiler set in ILogicalDevice

yes but to this one we don't have any access since it's protected, used only internally by `createShader` method

Reply:
> what's the rationale for unmounting the archive?

discussed on the discord thread, we will not unmount & let users do it instead if they want to